File: 127-word-ladder/127-word-ladder.scala

import scala.collection.mutable._

object Solution {
    def ladderLength(beginWord: String, endWord: String, wordList: List[String]): Int = {
        val graph: Map[String, Set[String]] = Map()

        (wordList :+ beginWord).foreach { w1 =>
            val set: Set[String] = Set()

            wordList.foreach { w2 =>
                if (w1 != w2 && isConnected(w1, w2)) {
                    set += w2
                }
            }

            graph += (w1 -> set)
        }

        val visited: Set[String] = Set()
        val queue: Queue[String] = Queue()

        queue.enqueue(beginWord)
        visited += beginWord

        var res = 0
        var found = false

        while(!found && queue.nonEmpty) {
            val size = queue.size

            for(i <- 1 to size) {
                val word = queue.dequeue

                if (word == endWord) { found = true }
                else {
                    val connected = graph.get(word).getOrElse(Set())
                    connected.foreach { other =>
                        if (!visited.contains(other)) {
                            queue.enqueue(other)
                            visited += other
                        }
                    }
                }
            }

            res += 1
        }

        if (found) res else 0
    }

    def isConnected(w1: String, w2: String): Boolean = {
        if (w1.length == w2.length) {
            var diff = 0

            for(i <- 0 to w1.length - 1) {
                diff += (if (w1.charAt(i) != w2.charAt(i)) 1 else 0)
            }

            diff == 1
        } else false
    }
}

File: 127-word-ladder/127-word-ladder.java
class Solution {
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        if (beginWord.equals(endWord)) return 1;
        
        Map<String, Set<String>> edges = new HashMap<>();
        
        for(int i = 0; i < wordList.size() - 1; i++) {
            String a = wordList.get(i);
            
            for(int j = i + 1; j < wordList.size(); j++) {
                String b = wordList.get(j);
                
                if (connected(a, b)) {
                    Set<String> s1 = edges.computeIfAbsent(a, k -> new HashSet<>());
                    Set<String> s2 = edges.computeIfAbsent(b, k -> new HashSet<>());
                    s1.add(b);
                    s2.add(a);
                }
            }
        }
        
        Queue<String> queue = new LinkedList<>();
        Set<String> visited = new HashSet<>();
        
        visited.add(beginWord);
        
        for(int i = 0; i < wordList.size(); i++) {
            String w = wordList.get(i);
            if (visited.contains(w)) continue;
            if (connected(beginWord, w)) {
                queue.offer(w);
                visited.add(w);
            }
        }
        
        int len = 2;
        
        while(!queue.isEmpty()) {
            int sz = queue.size();
            
            for(int i = 0; i < sz; i++) {
                String w = queue.poll();
                if (w.equals(endWord)) return len;
                
                for(String connect : edges.getOrDefault(w, new HashSet<>())) {
                    if (visited.contains(connect)) continue;
                    queue.offer(connect);
                    visited.add(connect);
                }
            }
            
            len++;
        }
            
        return 0;
    }
    
    boolean connected(String a, String b) {
        boolean flag = false;
        for(int i = 0; i < a.length(); i++) {
            if (a.charAt(i) != b.charAt(i)) {
                if (flag) return false;
                flag = true;
            }
        }
        return true;
    }
}